### PR TITLE
Holoperson Tweaks

### PR DESCRIPTION
**Pull request by Shadowmech88**
This doesn't re-add holopeople, yet.

* Number of active holopeople limited to 4.
* Holopeople can no longer spawn when the Holodeck control console is unpowered or broken.
* Holopeople now receive more specific rules regarding their behavior.
* Holopeople now receive names rather than numbers, to make reporting easier.
* Spawning as a holoperson is now done through the Holodeck control console interface.
* Holopeople who ghost will now be despawned instead of leaving behind a dumb holoperson mob.

### CODE REVIEW

**Comment by Shadowmech88** [comment]
Also, I need ideas for names with which to populate the names list.
Unfortunately there don't seem to be a whole lot of hologram characters in media.

**Comment by ShiftyRail** [comment]
Ideally holopeople should just do "holodamage" unless safety is off.
Holopeople should not have any mean to interact with the console.
A broken window should damage the holodeck and cause all hologhosts to disapear.

Holopeople freeing themselves and wrecking havoc _could_ be a fun bus, I guess.

**Comment by Durgen** [comment]
Good changes. Emagging the console should lift the cap on holopeople.

**Comment by ShiftyRail** [comment]
As for names, do we allow people to choose their own or would it end up with one billion "My body is on the Asteroi" clones ?

**Comment by Exxion** [comment]
Oh, also holopeople who ghost and re-hologram leave behind *ghosts*, as well
Someone made probably a hundred `ckey`less ghosts the last round they were in
Might want to fix that

**Comment by unid15** [comment]
I think players should be able to modify the max amount of holograms using the console. This way if holodudes are being shit, you can just set the max at 0 so they all disappear.

**Comment by Shadowmech88** [comment]
>holopeople who ghost and re-hologram leave behind ghosts

Where did you see these ghosts?
My *guess* is that the ghosts were in the contents of the dumb holomobs, which this will fix because those mobs now get deleted.

**Comment by Shadowmech88** [comment]
@unid15 I feel like that will just lead to everyone rushing the console at roundstart to set it to 0 for fear of the holograms.
I'd rather allow the holograms to always spawn so that the ones who break the rules get banned and people stop fearing the holograms.

**Comment by unid15** [comment]
Makes sense from a logical standpoint though, if the holograms are capable of killing somebody then there should be a way to turn them off. Perhaps the hologram controls could be password protected (with the password spawning at RD's).

**Comment by PJB3005** [comment]
Have a "kill all holopeople" button and a "disable holopeople spawn" button.

File: code/game/machinery/computer/HolodeckControl.dm [hunk 1]
```diff
@@ -22,22 +24,34 @@
...
-//		user.transmogrify(/mob/living/simple_animal/hologram/advanced, TRUE)
+/obj/machinery/computer/HolodeckControl/proc/spawn_holoperson(mob/dead/observer/user)
+	if(stat & (NOPOWER|BROKEN|MAINT))
+		return
+	if(!ticker || ticker.current_state != GAME_STATE_PLAYING)
+		to_chat(user, "<span class='notice'>You can't do this until the game has started.</span>")
+		return
+	if(!linkedholodeck)
+		return
+	if(connected_holopeople.len >=4)
```
Comment (PJB3005) [inline code/game/machinery/computer/HolodeckControl.dm]: You're not using the `maximum_holopeople` variable here.

File: code/game/machinery/computer/HolodeckControl.dm [hunk 2]
```diff
@@ -22,22 +24,34 @@
...
+		to_chat(user, "<span class='notice'>\The [src] cannot sustain any additional advanced holograms. Please try again when there are fewer advanced holograms on the holodeck.</span>")
+		return
+	var/turf/spawnturf
+	var/list/L = get_area_turfs(linkedholodeck.type)
+	L = shuffle(L)
```
Comment (PJB3005) [inline code/game/machinery/computer/HolodeckControl.dm]: Using a while loop with a pick is probably a better solution.

File: code/game/machinery/computer/HolodeckControl.dm [hunk 3]
```diff
@@ -46,6 +60,9 @@
 	user.set_machine(src)
 	var/dat
 
+	dat += {"<B>Holodeck Control System</B><BR>"}
```
Comment (PJB3005) [inline code/game/machinery/computer/HolodeckControl.dm]: I think you duplicated this line a bit.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/game/machinery/computer/HolodeckControl.dm (hunk 2 above)
Comment (Shadowmech88) [reply code/game/machinery/computer/HolodeckControl.dm]: Shuffling and then cycling through ensures both that a random turf is chosen and that no turf is checked twice. Using `pick()` in a loop could result in unnecessary duplicate checks.

Comment (PJB3005) [reply code/game/machinery/computer/HolodeckControl.dm]: The shuffle itself would FAR outweigh the performance lost from a few rare duplicates.

**Comment by Blithering** [comment]
Would it be possible to have an emergency hologram in medbay?

**Review by PJB3005 — APPROVED** [review]

**Comment by Shadowmech88** [comment]
This isn't ready yet, I still need ideas for names with which to populate the list of random hologram names.

**Comment by DrSnips** [comment]
Putting a few name suggestions here.
Gatebox, Rimmer, and Switch/Switchtool.

**Comment by Blithering** [comment]
Pun names!
Male: Hal O'Gram
Female: Holly Gram

**Comment by Durgen** [comment]
Neo, Morpheus, Trinity, Cypher, Apoc, Dozer, Tank, Switch, Mouse, Architect, Oracle, Smith

**Pull request merged**